File: app/lib/device/edge-switch.ts
import AsyncLock from "async-lock"
import { NodeSSH } from "node-ssh"
import { getAppConfig } from "../config/config"
import { parseTotalTransmit } from "./total-transmit-parser"

const lock = new AsyncLock({ timeout: 5000 })

const exec = async (commands: string[], config = getAppConfig().edgeswitch): Promise<string> => {
    const messages: string[] = []

    await lock.acquire<string>("ssh", async (done) => {
        const ssh = new NodeSSH()

        const client = await ssh.connect({
            host: config.ip,
            username: config.username,
            password: config.password
        })

        return new Promise((resolve) => {
            client.requestShell().then(async (shell) => {
                shell.on("data", (data: Buffer) => {
                    messages.push(data.toString("utf8"))
                })
                shell.on("close", async () => {
                    resolve(messages.join(""))
                    done()
                })

                for (const command of commands) {
                    await shell.write(`${command}\n`)
                }
                await shell.end()
            })
        })
    })

    return messages.join("")
}

export const turnOff = async (port: string) => {
    await exec([
        "configure",
        `interface ${port}`,
        "poe opmode shutdown"]
    )
}

export const turnOn = async (port: string) => {
    await exec([
        "configure",
        `interface ${port}`,
        "poe opmode auto"]
    )
}

export const status = async (port: string) => {
    const message = await exec([
        "configure",
        `show poe status ${port}`]
    )

    const result = (message ?? "")
        .split("\n")
        .filter((line) => line.startsWith(port))
        .map(line => line.trim())
        .map((line) => line.split(/\s+/g))[0]

    if (!result || result.length === 0) {
        return undefined
    }

    return {
        interface: result[0],
        detection: result[1],
        status: result[1].toLowerCase() === "good" ? 1 : 0,
        class: result[2],
        energy: +result[3],
        voltage: +result[4],
        current_mA: +result[5],
        total_Whr: +result[6],
        temperature: +result[7]
    }
}

export const statusTotalTransmit = async () => {
    const message= await exec([
        "configure",
        `show interface ethernet all`]
    )

    return parseTotalTransmit(message)
}

File: app/edgeswitch/edgeswitch.go
package edgeswitch

import (
    "rnd7/edgeswitch-mqtt/config"
    "rnd7/edgeswitch-mqtt/logger"
    "rnd7/edgeswitch-mqtt/mqtt"
    "strings"
)

var cfg config.Config
var portToName = make(map[string]string)
var nameToPort map[string]string

func Configure(configuration config.Config) {
    cfg = configuration

    for _, portInfo := range cfg.EdgeSwitch.Ports {
        portToName[portInfo.Port] = portInfo.Name
    }

    nameToPort = make(map[string]string)
    for _, portInfo := range configuration.EdgeSwitch.Ports {
        nameToPort[portInfo.Name] = portInfo.Port
    }
}

func OnMessage(topic string, payloadData []byte) {
    payload := string(payloadData)
    paths := strings.Split(topic, "/")
    name := strings.TrimSpace(paths[len(paths) - 3])
    port := nameToPort[name]

    if port == "" {
        logger.Error("Port not found for name", name)
        return
    }

    if strings.EqualFold(payload, "true") {
        TurnPoEOn(port)
    } else if strings.EqualFold(payload, "false") {
        TurnPoEOff(port)
    }
}

func TurnPoEOff(port string) {
    logger.Info("Switching PoE OFF for port", port)
    poeSwitch(port, "shutdown")
}

func TurnPoEOn(port string) {
    logger.Info("Switching PoE ON for port", port)
    poeSwitch(port, "auto")
}

func poeSwitch(port string, opmode string) {
    execute([]string{
        "configure",
        "interface " + port,
        "poe opmode " + opmode})
}

func execute(commands []string) {
    logger.Debug("Executing commands", commands)
    session, err := StartSession(cfg.EdgeSwitch.Username, cfg.EdgeSwitch.Password, cfg.EdgeSwitch.IP + ":22")
    if err != nil {
        logger.Error("StartSession failed", err)
        return
    }
    logger.Debug("Session started")

    defer session.Close()

    for _, cmd := range commands {
        session.Write(cmd)
        var x = session.ReadChannelData()
        logger.Debug("Command:", cmd, "Response:", x)
    }
}

func Execute() {
    logger.Debug("Updating...")

    session, err := StartSession(cfg.EdgeSwitch.Username, cfg.EdgeSwitch.Password, cfg.EdgeSwitch.IP + ":22")
    if err != nil {
        logger.Error("StartSession failed", err)
        return
    }

    defer session.Close()

    // Disable pagination
    session.Write("terminal length 0")
    var _ = session.ReadChannelData()

    session.Write("configure")
    var _ = session.ReadChannelData()

    session.Write("show interface ethernet all")
    channelData, err := ParseChannelData(session.ReadChannelData())
    if err != nil {
        return
    }

    for _, chdata := range channelData {
        mqtt.PublishJSON("ports/" + toName(portToName, chdata.Port) + "/transmit", mqtt.TransmitMessage{
            Port: chdata.Port,
            BytesTx: chdata.BytesTx,
            BytesRx: chdata.BytesRx,
            PacketsTx: chdata.PacketsTx,
            PacketsRx: chdata.PacketsRx,
            TotalBytes: chdata.BytesTx + chdata.BytesRx,
        })
    }

    aggregated := mqtt.AggregatedEnergy{
        EnergySum: 0.0,
        WhrSum: 0.0,
    }

    for _, port := range cfg.EdgeSwitch.Ports {
        session.Write("show poe status " + port.Port)
        info, err := ParseDeviceInfo(session.ReadChannelData())
        if err != nil {
            return
        }

        for _, deviceInfo := range info {
            mqtt.PublishJSON("ports/" + toName(portToName, deviceInfo.Intf) + "/poe", mqtt.DeviceDataMessage{
                Interface:   deviceInfo.Intf,
                Detection:   deviceInfo.Detection,
                Status:      convertStatus(deviceInfo.Detection),
                Class:       deviceInfo.Class,
                Energy:      deviceInfo.ConsumedW,
                Voltage:     deviceInfo.VoltageV,
                CurrentmA:   deviceInfo.CurrentmA,
                TotalWhr:    deviceInfo.ConsumedMeterWhr,
                Temperature: deviceInfo.TemperatureC,
            })

            aggregated.EnergySum += deviceInfo.ConsumedW
            aggregated.WhrSum += deviceInfo.ConsumedMeterWhr
        }
    }

    // round to 2 decimal places
    aggregated.EnergySum = float64(int(aggregated.EnergySum * 100)) / 100
    aggregated.WhrSum = float64(int(aggregated.WhrSum * 100)) / 100

    mqtt.PublishJSON("aggregated", aggregated)

    logger.Debug("Update completed")
}


func convertStatus(statusString string) int {
    // Compare the string case-insensitively
    if strings.EqualFold(statusString, "Good") {
        return 1
    } else {
        return 0
    }
}

func toName(portToName map[string]string, port string) string {
    name := portToName[port]
    if name == "" {
        name = port
        name = strings.ReplaceAll(name, "/", "_")
    }
    return name

}
